kgraphql/src/main/kotlin/com/apurebase/kgraphql/schema/SchemaPrinter.kt

```diff
@@ -0,0 +1,290 @@
+                        appendDescription(field, indentation)
+                        appendLine("${indentation}${field.name}: ${field.type.typeReference()}${field.deprecationInfo()}")
+                    }
+                    appendLine("}")
+                }
+        }
+
+        // Directives
+        //   https://spec.graphql.org/draft/#sec-Type-System.Directives.Built-in-Directives
+        //   "When representing a GraphQL schema using the type system definition language any built-in directive may be omitted for brevity."
+        val directives = buildString {
+            schema.directives.filter { config.includeBuiltInDirectives || !it.isBuiltIn() }.sortedByName()
+                .forEachIndexed { index, directive ->
+                    if (index > 0) {
+                        appendLine()
+                    }
+                    val args = directive.args.takeIf { it.isNotEmpty() }
+                        ?.joinToString(", ", prefix = "(", postfix = ")") { arg ->
+                            arg.name + ": " + arg.type.typeReference() + arg.defaultInfo()
+                        } ?: ""
+                    appendLine("directive @${directive.name}$args on ${directive.locations.joinToString(" | ") { it.name }}")
+                }
+        }
+
+        return listOf(
+            schemaDefinition,
+            scalars,
+            objects,
+            interfaces,
+            unions,
+            enums,
+            inputTypes,
+            directives
+        ).filterNot { it.isBlank() }.joinToString("\n")
+    }
+
+    private fun includeSchemaDefinition(schema: __Schema): Boolean =
+        config.includeSchemaDefinition ||
+            schema.queryType.name != "Query" ||
+            (schema.mutationType != null && schema.mutationType?.name != "Mutation") ||
+            (schema.subscriptionType != null && schema.subscriptionType?.name != "Subscription") ||
+            schema.types.any {
+                it != schema.queryType && it.name == "Query" ||
+                    it != schema.mutationType && it.name == "Mutation" ||
+                    it != schema.subscriptionType && it.name == "Subscription"
+            }
+
+    private fun Depreciable.deprecationInfo(): String = if (isDeprecated) {
+        " @deprecated(reason: \"$deprecationReason\")"
+    } else {
+        ""
+    }
+
+    private fun __Directive.isBuiltIn(): Boolean =
+        name in setOf(Directive.DEPRECATED.name, Directive.INCLUDE.name, Directive.SKIP.name)
+
+    private fun __Type.isBuiltInScalar(): Boolean = name in BuiltInScalars.entries.map { it.typeDef.name }
+
+    private fun __Type.implements(): String =
+        interfaces
+            .sortedByName()
+            .mapNotNull { it.name }
+            .takeIf { it.isNotEmpty() }
+            ?.joinToString(separator = " & ", prefix = " implements ") ?: ""
+
+    private fun Any.description(): List<String>? = when (this) {
```

Comment:
Hmmm, this looks like we are missing an interface like `Describable` or something like that, that all describable types are implementing.

Reply:
Actually there is a `__Described` interface but `__Type` does not implement it. Likely due to a conflict with the non-optional `name`. Since this is the only place where I needed it I decided to postpone looking into that. Maybe there is another need for it when implementing https://github.com/stuebingerb/KGraphQL/issues/94